lisa/tests/kernel/scheduler/load_tracking.py

```diff
@@ -403,6 +403,20 @@ def test_task_load_avg(self, allowed_error_pct=15) -> ResultBundle:
         """
         return self._test_signal('load_avg', allowed_error_pct)
 
+
+    def test_slack(self, negative_slack_allowed_pct=15) -> ResultBundle:
+        """
+        Same test as :meth:`RTATestBundle.test_slack`, but executed on all
+        the collected traces.
+        """
+        return CombinedResultBundle(
```

Comment:
Ah right, this one needs special attention here.

For now I'd be inclined not to introduce a new `ResultBundle` class. `TestMetric`s can be nested inside a `ResultBundle`, so I would see this as being the same as `RTATestBundle.test_slack` but with an extra metric level (per frequency).


Adding this class could make things more confusing for new users. If a single test depends on several metrics (e.g. `test_slack()` has per-task slack score), what should be used? `CombinedResultBundle`s with `ResultBundle` per task "test"? `ResultBundle` with several metrics?

I'd like to keep things as simple as possible and stick with `ResultBundle`+`TestMetric`. It just so happens here that we're basically doing n iterations of `test_slack()`, so it looks like it makes sense to create a class that handles "iterations of `ResultBundle`s", but this frequency test is an ugly duckling that doesn't fit our usual test pattern (a single trace per workload).

Reply:
The thing is that if you have a class that provides the `from_testenv()` infra for rt-app, you have everything that's required to run `test_slack()` (running an rt-app workload generates the file required for this test). It might make more sense to override `test_slack()` with just an exception raise (and you can strip out the annotations for exekall).

Then I don't know how the whole `TestBundle` logic would work - you only have one `TestBundle` for the frequency test, but you also get several `TestBundle`s for the slack test?